File: medium/0008_string_to_integer/solution.cpp
#include <string>
#include <vector>

class Solution {
public:
  int myAtoi(std::string s) {
    const int sLen = s.size();

    int i = 0;
    while (i < sLen && s[i] == ' ')
      ++i;

    int sign = 1;
    if (s[i] == '+' || s[i] == '-') {
      sign = s[i] == '-' ? -1 : 1;
      ++i;
    }

    long long num = 0;

    while (i < sLen && isdigit(s[i])) {
      num = num * 10 + (s[i++] - '0');

      if (sign == 1 && num > INT_MAX)
        return INT_MAX;
      if (sign == -1 && -num < INT_MIN)
        return INT_MIN;
    }

    return (int)(sign * num);
  }
};

File: medium/0008_string_to_integer/solution.py
class Solution:
    def myAtoi(self, s: str) -> int:
        s_len = len(s)

        i = 0
        while i < s_len and s[i] == " ":
            i += 1

        sign = 1
        if i < s_len and (s[i] == "+" or s[i] == "-"):
            sign = -1 if s[i] == "-" else 1
            i += 1

        num = 0

        while i < s_len and s[i].isdigit():
            num = num * 10 + (ord(s[i]) - 48)

            if sign == 1 and num > 2147483647:
                return 2147483647
            if sign == -1 and -num < -2147483648:
                return -2147483648

            i += 1

        return sign * num
